Added verification script for Swagger API object descriptions

stevekuznetsov opened the pull request:
Begins to fix #3830 
Need help describing these objects.

@deads2k fyi

deads2k commented on hack/verify-api-descriptions.sh:
```diff
@@ -0,0 +1,51 @@
+cd "${OS_ROOT}"
+
+DEF_MISSING=false
+TYPES_LOCATIONS=$(grep -Po "(?<=\_ \")[A-Za-z0-9\.\/]+(?=\")" pkg/api/${1}/register.go)
```
Not all types are directly used by the user and grepping through a sort file is nasty.  Try walking the swagger.  We'll have to do that for other work anyway.

stevekuznetsov commented:
@deads2k Updated to walk Swagger spec for types and reference whitelist.

stevekuznetsov commented:
@liggitt [Panic occurring in Travis](https://travis-ci.org/openshift/origin/jobs/72185509#L6936-6940) due to `ldap.go`:
```
I0722 21:46:17.499379   16980 ldap.go:193] no entries matching (&(myAttr=myValue)myuid1053a8ff-30bb-11e5-9c11-f2caf0029c69=myuser))2015/07/22 21:46:17 Eror readMessagePacket: EOF
2015/07/22 21:46:17 client 1 close()
2015/07/22 21:46:17 Client [1] sent abandon signal to request[messageID = 2]
panic: close of nil channel
```

stevekuznetsov commented:
[test]

stevekuznetsov commented:
@deads2k  PTAL

deads2k commented on hack/verify-api-descriptions.sh:
```diff
@@ -0,0 +1,48 @@
+OS_ROOT=$(dirname "${BASH_SOURCE}")/..
+source "${OS_ROOT}/hack/common.sh"
+
+cd "${OS_ROOT}"
```
Do we do this elsewhere?  I'm not a huge fan of changing directories in the shell process.

stevekuznetsov replied:
That's in the standard preamble in the following scripts: 
```
hack/build-base-images.sh
hack/build-images.sh
hack/build-release.sh
hack/common.sh
hack/copy-kube-artifacts.sh
hack/push-release.sh
hack/release.sh
hack/test-assets.sh
hack/test-extended.sh
hack/test-go.sh
hack/test-integration-docker.sh
hack/test-integration.sh
hack/update-generated-conversions.sh
hack/update-generated-deep-copies.sh
hack/verify-generated-completions.sh
hack/verify-generated-conversions.sh
hack/verify-generated-deep-copies.sh
hack/verify-generated-docs.sh
hack/verify-generated-swagger-spec.sh
hack/verify-gofmt.sh
hack/verify-golint.sh
hack/verify-jsonformat.sh
```

deads2k commented on hack/verify-api-descriptions.sh:
```diff
@@ -0,0 +1,48 @@
+set -o pipefail
+
+if [ "$#" -lt 1 ]; then
+	echo "Usage: verify-api-descriptions.sh <OpenShift API version>"
```
Seems like this ought to inspect all versions by default.

deads2k commented on hack/verify-api-descriptions.sh:
```diff
@@ -0,0 +1,48 @@
+	then
+		if [ ! -s "${desc_location}/description.adoc" ]
+		then
+			if ! grep -q "${object}" hack/api-description-whitelist.txt
```
check for start and end of line to avoid getting partial matches.

deads2k commented on hack/api-description-whitelist.txt:
```diff
@@ -0,0 +1,150 @@
+json.watchevent
+patch.object
+runtime.rawextension
+v1.awselasticblockstorevolumesource
```
@smarterclayton check + whitelist while we work the backlog seems reasonable.  Are there any we want to permanently whitelist?  Also, is this roughly the list you expect?

smarterclayton replied:
Yeah this is fine.

smarterclayton commented on hack/list-swagger-objects.py:
```diff
@@ -0,0 +1,13 @@
+import json;
```
Is this the code upstream has?

stevekuznetsov replied:
Upstream does not have this code. Upstream `grep`s through `register.go` to find everything that has an `IsAnAPIObject` method. This is not the same, as the items that Swagger picks up.

deads2k commented on hack/verify-api-descriptions.sh:
```diff
@@ -0,0 +1,59 @@
+	fi
+}
+
+VERSIONS="v1beta3 v1"
```
Put v1beta3 in a blacklist and find a way to auto-detect the versions

stevekuznetsov commented:
@pmorie Travis [fail](https://travis-ci.org/openshift/origin/jobs/73028024#L1458) on Godeps test:
```
--- FAIL: TestRequestExecuteRemoteCommand-2 (0.00s)
	remotecommand_test.go:207: 1: expected true, got false: <nil>
FAIL
coverage: 80.8% of statements
FAIL	github.com/openshift/origin/Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client/remotecommand	0.086s
```

stevekuznetsov commented:
@deads2k Made the following changes:
* added logic to grab all versions of swagger spec from the server in `hack/update-generated-swagger-spec.sh`
* added logic to handle all available swagger specs in `hack/verify-api-descriptions.sh`
* added `hack/prune-api-description-whitelist.sh` to remove unnecessary whitelist entries
* added `hack/propagate-api-descriptions.sh` to take description files in `api/definitions/raw` and propagate them to all versions they need to support (for instance, `node.adoc` would become the `description.adoc` for `v1beta3.node` as well as `v1.node`. I chose not to allow drifting swagger spec between versions because that's a logistical nightmare to deal with. @smarterclayton  if you think we need it please say so. otherwise we just need to make sure that swagger descriptions are relatively version-agnostic

deads2k commented on hack/api-description-whitelist.txt:
```diff
@@ -0,0 +1,329 @@
+patch.object
+runtime.rawextension
+v1.awselasticblockstorevolumesource
+v1beta3.awselasticblockstorevolumesource
```
Any idea how or why you're picking up v1beta3?  We don't have a generated swagger spec for that one

deads2k commented on hack/prune-api-description-whitelist.sh:
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
```
I don't think I care about this script.  I'm inclined to make people keep up with it manually.  You have good messages in the verify script already.

stevekuznetsov replied:
Is it wrong to have? I'm going to be writing these things and it's pretty useful I will admit ...

deads2k replied:
> Is it wrong to have? I'm going to be writing these things and it's pretty useful I will admit ...

It's one more thing to rot.  If you've gone to the trouble of writing a description, deleting a whitelist entry is minimal effort.  Especially since you have the correct messages in the verify script.

deads2k commented on hack/propagate-api-descriptions.sh:
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
```
Unless we're checking that they match (doubtful and probably incorrect), I'd say the verify script is sufficient to smack people.  This saves them a copy/paste at the cost of them having to know to use it.

deads2k commented on hack/update-generated-swagger-spec.sh:
```diff
@@ -61,8 +61,17 @@ popd > /dev/null
+done
+
+KAPI_ENDPOINTS=$(curl "${HOST}" | grep -Po "(?<=\/api\/)[a-z0-9]+")
+for endpoint in $KAPI_ENDPOINTS
```
blacklist v1beta3

stevekuznetsov commented:
@deads2k added `v1beta3` blacklist

deads2k commented on api/definitions/.gitignore:
```diff
@@ -0,0 +1 @@
+raw/
```
huh?

deads2k commented on api/definitions/raw/buildconfig.adoc:
```diff
@@ -0,0 +1,3 @@
+Build configurations define a build process for new Docker images. There are three types of builds possible - a Docker build using a Dockerfile, a Source-to-Image build that uses a specially prepared base image that accepts source code that it can make runnable, and a custom build that can run arbitrary Docker images as a base and accept the build parameters. Builds run on the cluster and on comp ...
```
I think it's best to keep these under v1

deads2k replied:
These should be gone, right?

deads2k commented:
remove dead files and unnecessary scripts and I think this is good to go.

stevekuznetsov commented:
@deads2k pruning done

deads2k commented on hack/update-generated-swagger-spec.sh:
```diff
@@ -61,8 +61,17 @@ popd > /dev/null
-curl "${SWAGGER_API_PATH}oapi/v1" > "${SWAGGER_SPEC_OUT_DIR}/oapi-v1.json"
-echo "Updating ${SWAGGER_SPEC_OUT_DIR}/api-v1.json from ${SWAGGER_API_PATH}api/v1..."
-curl "${SWAGGER_API_PATH}api/v1" > "${SWAGGER_SPEC_OUT_DIR}/api-v1.json"
+OAPI_ENDPOINTS=$(curl "${HOST}" | grep -Po "(?<=\/oapi\/)[a-z0-9]+" | sed '/v1beta3/d')
```
seems like api and oapi should be passed to a loop?  These are the same other than that, right?

deads2k commented:
lgtm [merge]
